### DISCUSSION TITLE
Using GPU Resources for Lava Simulations & Drastic Slow-Down when Probing

### DISCUSSION THREAD

**Discussion by avitaleSon** [lava-nc/lava#262] [Q&A]
Hello,

firstly: thanks a lot for open-sourcing the whole Lava framework! This is quite a powerful set of tools which will surely contribute to advancing the frontier of Neuromorphic Research!

I have two questions regarding the runtime performance of small workloads simulated with Lava:

1. Using GPU Resources:
  I have a question regarding running Lava Simulations on GPU. I am running a small workload on my workstation. I have a GPU enabled workstation, yet when I include the ```@requries(GPU)``` flag the runtime stays exactly the same as using ```@requires(CPU)```. Is there something that I am missing?
  ```python
  from lava.magma.core.resources import GPU
from lava.magma.core.sync.protocols.loihi_protocol import LoihiProtocol

class var_logger(AbstractProcess):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        shape = kwargs.get("shape", (1,))
        self.ref = RefPort(shape=shape)

  @implements(proc=var_logger, protocol=LoihiProtocol)
  @requires(GPU)
  class PyProcModel1(PyLoihiProcessModel):
  ...
  ```
2. Drastic Slow-Down when Probing:
In order to monitor the spiking activity of a population ```lif_1```, I am connecting it's output port to the input port of a ```io.sink.
RingBuffer``` in the following way:
```python
    lif1 = LIF(shape=shape, v=lif1_params['v'],
               du=lif1_params['du'], dv=lif1_params['dv'],
               bias=lif1_params['bias'], vth=lif1_params['vth'])

    lif_spike_logger = io.sink.RingBuffer(shape=lif1.shape, buffer=num_steps)
    lif1.s_out.connect(lif1_spike_logger.a_in)

    lif1.run(condition=RunSteps(num_steps=num_steps, blocking=True), run_cfg=Loihi1SimCfg()
```
When I include the RingBuffer probing object, the workload runtime increases by a factor of ~10 (without probing: 34s with probing: 320s). Is there a way to make this more efficient? For example, probing only every N-th timestep? What about implementing the probing in a Lava-independent way e.g. by saving the spike-timestamps and spike-NeuronIDs in a python list. Would this bring speed up benefits?

Thanks for your help!
Antonio

**Comment by tim-shea** [comment]
1. The @requires tag only specifies that the associated proc model ought to be run on a corresponding node, it does not automagically convert the guts of that proc model to use the right hardware. Thus, if you want PyProcModel1 to use a GPU, you would implement a GPU-based PyProc inside of it, provide the appropriate @requires tag, *and* ensure that your run_config results in a node that actually can provide the appropriate hardware at run time. As far as I know, there are no good examples of this for GPUs right now.
2. This performance issue may have been addressed in v0.4.0. Much of the python communication infrastructure was overhauled in this major update. Please let me know if you're still seeing a big performance penalty for this use case and we can submit an issue to try to chase down the problem.

**Reply by avitaleSon** [reply]
Hi Tim,

thanks a lot for your answers!

1. So basically in order to run the network simulation on GPU one has to manually define the process which uses GPU resources. Lava does not take care of this under the hood. Is there a plan to release a tutorial on how to implement Lava processes to run on different Hardware?

2. I will update my Lava release and test again the runtime performance. I will let you know as soon as I do this!

Antonio
